### PR TITLE
feat(manager/scalafmt): implement manager

### LINKED ISSUE #13877: Update version in .scalafmt.conf file
### What would you like Renovate to be able to do?

Renovate should be able to update the `version` field inside the `.scalafmt.conf` file.

### If you have any ideas on how this should be implemented, please tell us here.

I suppose it should be easy to implement custom file handling.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by gaeljw**
## Changes

This PR adds support for updating scalafmt version (https://github.com/renovatebot/renovate/issues/13877).

It does NOT add support for automatically reformating the codebase using the new scalafmt version as suggested in the issue, this could be implemented in a second step.

## Context

- Closes https://github.com/renovatebot/renovate/issues/13877

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - Real repo PRs: https://github.com/gaeljw/renovate_13877/pull/1, https://github.com/gaeljw/renovate_13877/pull/2, https://github.com/gaeljw/renovate_13877/pull/3

### CODE REVIEW

File: lib/modules/manager/scalafmt/extract.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+);
+
+export function extractPackageFile(
+  content: string,
+  packageFile: string,
+): PackageFileContent | null {
+  const regexResult = scalafmtVersionRegex.exec(content);
+  const scalafmtVersion = regexResult?.groups?.version;
+  const matchString = regexResult?.[0];
+  if (scalafmtVersion) {
```
Comment (secustor) [inline lib/modules/manager/scalafmt/extract.ts:24]: Reverse with early return to simplify.

File: lib/modules/manager/scalafmt/extract.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    const scalafmtDependency: PackageDependency = {
+      datasource: GithubReleasesDatasource.id,
+      depName: 'scalameta/scalafmt',
+      packageName: 'scalameta/scalafmt',
+      versioning: semverVersioning.id,
+      currentValue: scalafmtVersion,
+      replaceString: matchString,
```
Comment (secustor) [inline lib/modules/manager/scalafmt/extract.ts:31]:
```suggestion
```
Shouldn't be necessary

File: lib/modules/manager/scalafmt/extract.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      extractVersion: '^v(?<version>\\S+)',
+      registryUrls: [],
```
Comment (secustor) [inline lib/modules/manager/scalafmt/extract.ts:33]:
```suggestion
```
noop

File: lib/modules/manager/scalafmt/extract.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    const pkg = extractPackageFile(content, packageFile);
+    if (pkg) {
+      packages.push({ deps: pkg.deps, packageFile });
+    }
+  }
+
+  return packages;
+}
```
Comment (secustor) [inline lib/modules/manager/scalafmt/extract.ts:64]:
```suggestion
```
`extractAllPackageFiles` has only to be implement if all files matched by `fileMatch` in `index.ts` need to processed in the same context. 
https://github.com/renovatebot/renovate/blob/main/docs/development/adding-a-package-manager.md#extractallpackagefilespackagefiles-async-optional

File: lib/modules/manager/scalafmt/index.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import type { Category } from '../../../constants';
+import { GithubReleasesDatasource } from '../../datasource/github-releases';
+import * as semverVersioning from '../../versioning/semver';
+
+export { extractAllPackageFiles, extractPackageFile } from './extract';
+
+export const supportedDatasources = [GithubReleasesDatasource.id];
+
+export const defaultConfig = {
+  fileMatch: ['\\.scalafmt.conf$'],
```
Comment (secustor) [inline lib/modules/manager/scalafmt/index.ts:10]: Which names can these  `.scalafmt.conf` file have.  E.g. could it be `test.scalafmt.conf`?

If no, then it should be changed to: 
```suggestion
  fileMatch: ['(^|/)\\.scalafmt.conf$'],
```

File: lib/modules/manager/scalafmt/extract.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  packageFile: string,
+): PackageFileContent | null {
+  const regexResult = scalafmtVersionRegex.exec(content);
+  const scalafmtVersion = regexResult?.groups?.version;
+  const matchString = regexResult?.[0];
+  if (scalafmtVersion) {
+    const scalafmtDependency: PackageDependency = {
+      datasource: GithubReleasesDatasource.id,
+      depName: 'scalameta/scalafmt',
+      packageName: 'scalameta/scalafmt',
```
Comment (secustor) [inline lib/modules/manager/scalafmt/extract.ts:28]:
```suggestion
```
Only one of `depName` or `packageName` are needed if they contain the same value. 
The first should be human-readable and the later is the name which is actually looked up by the datasource.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/scalafmt/extract.ts (hunk 6 above)
Comment (gaeljw) [reply lib/modules/manager/scalafmt/extract.ts]: I kept "scalafmt" as human-readable name then, that's how most people know the tool.

**Comment by gaeljw** [comment]
Thanks for the review @secustor :pray:

File: lib/modules/manager/scalafmt/index.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import { GithubReleasesDatasource } from '../../datasource/github-releases';
+import * as semverVersioning from '../../versioning/semver';
+
+export { extractPackageFile } from './extract';
+
+export const supportedDatasources = [GithubReleasesDatasource.id];
+
+export const defaultConfig = {
+  fileMatch: ['(^|/)\\.scalafmt.conf$'],
+  versioning: semverVersioning.id,
```
Comment (secustor) [inline lib/modules/manager/scalafmt/index.ts:11]:
```suggestion
```
Overlooked that one.

**Review by secustor — CHANGES_REQUESTED** [review]
Else lgtm

**Review by secustor — APPROVED** [review]

**Pull request merged**